### PR TITLE
[Checkout] Preventing payment step completion without a selected method  

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 1]
```diff
@@ -1212,6 +1212,36 @@ public function thereShouldBeInformationAboutNoShippingMethodsAvailableFormMyShi
...
+        // Intentionally left blank to fulfill context expectation
+    }
+
+    /**
+     * @Then I should not be able to complete the payment step
+     */
+    public function iShouldNotBeAbleToCompleteThePaymentStep()
+    {
+        Assert::true(
+            $this->selectPaymentPage->hasNoAvailableNextStepButton(),
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: How about `isNextStepButtonUnavailable`?

File: src/Sylius/Bundle/ShopBundle/Resources/views/Checkout/SelectPayment/_no_payment_methods_available.twig [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<div class="ui icon message" id="sylius-order-cannot-be-paid">
+    <i class="warning sign icon"></i>
+    <div class="content">
+        <div class="header">{{ 'sylius.ui.warning'|trans }}</div>
+        <p>{{ 'sylius.ui.no_payment_methods_available'|trans }}.</p>
+    </div>
+</div>
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Checkout/SelectPayment/_no_payment_methods_available.twig]: Missing blank line

File: src/Sylius/Component/Core/Resolver/DefaultPaymentMethodResolver.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefaultPaymentMethod(PaymentInterface $subject)
+    {
+        /** @var CorePaymentInterface $shipment */
+        Assert::isInstanceOf($shipment, CorePaymentInterface::class);
+
+        $paymentMethods = $this->paymentMethodRepository->findBy(['enabled' => true]);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Resolver/DefaultPaymentMethodResolver.php]: We should use `$this->paymentMethodRepository->findEnabledForChannel($channel)` method.

File: src/Sylius/Component/Core/Resolver/DefaultPaymentMethodResolver.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        /** @var ChannelInterface $channel */
+        $channel = $subject->getOrder()->getChannel();
+
+        foreach ($paymentMethods as $paymentMethod) {
+            if ($channel->hasPaymentMethod($paymentMethod)) {
+                return $paymentMethod;
+            }
+        }
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Resolver/DefaultPaymentMethodResolver.php]: This logic will not be necessary after the above comment, but should be removed from the `DefaultShippingMethodResolver` also (in a separate PR).

File: features/checkout/paying_for_order/preventing_payment_step_completion_without_a_selected_method.feature [hunk 5]
```diff
@@ -0,0 +1,53 @@
+@checkout
+Feature: Preventing payment step completion without a selected method
+    In order to be prevent from finish payment step without selected method
```
Comment (GSadee) [inline features/checkout/paying_for_order/preventing_payment_step_completion_without_a_selected_method.feature]: I guess it should be **prevented**

File: features/checkout/paying_for_order/preventing_payment_step_completion_without_a_selected_method.feature [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    As a Customer
+    I want to be prevent from finish payment step without selected method
```
Comment (GSadee) [inline features/checkout/paying_for_order/preventing_payment_step_completion_without_a_selected_method.feature]: As above

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 7]
```diff
@@ -1212,6 +1212,36 @@ public function thereShouldBeInformationAboutNoShippingMethodsAvailableFormMyShi
     }
 
     /**
+     * @Given I do not select any payment method
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: `@When`

File: features/checkout/paying_for_order/preventing_payment_step_completion_without_a_selected_method.feature (hunk 5 above)
Comment (NeverResponse) [inline features/checkout/paying_for_order/preventing_payment_step_completion_without_a_selected_method.feature]: the above + `from finishing the payment step without selecting a method` / `from finishing the payment step with no method selected`

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 8]
```diff
@@ -1223,6 +1223,36 @@ public function iShouldNotBeAbleToCompleteTheShippingStep()
...
+    }
+
+    /**
+     * @Then I should not be able to complete the payment step
+     */
+    public function iShouldNotBeAbleToCompleteThePaymentStep()
+    {
+        Assert::true(
+            $this->selectPaymentPage->isNextStepButtonUnavailable(),
+           'The "next step" button should be disabled, but it does not.'
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: `but it is not.` although I'd be in favour of dropping this part in every method as it doesn't bring any value. If the message is in an exception and it says `The "next step" button should be disabled.` it's quite obvious what happened without an overly explicit explanation.

File: src/Sylius/Bundle/PaymentBundle/Resources/translations/validators.en.yml [hunk 9]
```diff
@@ -19,6 +19,8 @@ sylius:
     payment:
         currency_code:
             not_valid: The currency code you entered is invalid.
+        method:
+            not_blank: 'Please select payment method.'
```
Comment (NeverResponse) [inline src/Sylius/Bundle/PaymentBundle/Resources/translations/validators.en.yml]: `a payment method.`

File: src/Sylius/Bundle/ShopBundle/Resources/views/Checkout/SelectPayment/_form.html.twig [hunk 10]
```diff
@@ -22,15 +23,25 @@
...
-                            <div class="header">{{ 'sylius.ui.warning'|trans }}</div>
-                            <p>{{ 'sylius.ui.no_payment_methods_available'|trans }}.</p>
-                        </div>
-                    </div>
+                    {% set disabled = true %}
+                    {% include '@SyliusShop/Checkout/SelectPayment/_no_payment_methods_available.twig' %}
                 {% endfor %}
             </div>
         </div>
+    {% else %}
```
Comment (NeverResponse) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Checkout/SelectPayment/_form.html.twig]: Is it even physically possible to get to this checkout step without a payment?

File: src/Sylius/Component/Payment/spec/Exception/UnresolvedDefaultPaymentMethodExceptionSpec.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+use Sylius\Component\Payment\Exception\UnresolvedDefaultPaymentMethodException;
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class UnresolvedDefaultPaymentMethodExceptionSpec extends ObjectBehavior
+{
+
```
Comment (NeverResponse) [inline src/Sylius/Component/Payment/spec/Exception/UnresolvedDefaultPaymentMethodExceptionSpec.php]: blank line

File: src/Sylius/Component/Payment/spec/Resolver/DefaultPaymentMethodResolverSpec.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        PaymentMethodInterface $secondPaymentMethod,
+        PaymentMethodRepositoryInterface $paymentMethodRepository,
+        PaymentInterface $payment
+    ) {
+        $paymentMethodRepository->findBy(['enabled' => true])->willReturn([$firstPaymentMethod, $secondPaymentMethod]);
+
+        $this->getDefaultPaymentMethod($payment)->shouldReturn($firstPaymentMethod);
+    }
+
+    function it_throws_exception_if_there_is_no_enabled_payment_methods(
```
Comment (NeverResponse) [inline src/Sylius/Component/Payment/spec/Resolver/DefaultPaymentMethodResolverSpec.php]: `there_are_no`

File: src/Sylius/Component/Payment/spec/Resolver/DefaultPaymentMethodResolverSpec.php [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace spec\Sylius\Component\Payment\Resolver;
+
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
```
Comment (NeverResponse) [inline src/Sylius/Component/Payment/spec/Resolver/DefaultPaymentMethodResolverSpec.php]: unused

File: src/Sylius/Bundle/PaymentBundle/Resources/translations/validators.en.yml [hunk 14]
```diff
@@ -36,4 +38,4 @@ sylius:
                 not_blank: Please enter the fee percent.
         code:
             not_blank: 'Please enter payment method code.'
-            unique: 'The payment method with given code already exists.'
+            unique: 'The payment method with given code already exists.'
```
Comment (NeverResponse) [inline src/Sylius/Bundle/PaymentBundle/Resources/translations/validators.en.yml]: Missing blank line.

File: tests/Controller/CheckoutPaymentApiTest.php [hunk 15]
```diff
@@ -54,7 +54,7 @@ public function it_does_not_allow_to_select_payment_for_order_that_is_not_addres
         $this->addressOrder($orderId);
 
         $url = sprintf('/api/checkouts/select-payment/%d', $orderId);
-        $this->client->request('PUT', $url, [], [], static::$authorizedHeaderWithContentType);
+        $this->client->request('PATCH', $url, [], [], static::$authorizedHeaderWithContentType);
```
Comment (tuka217) [inline tests/Controller/CheckoutPaymentApiTest.php]: screen shot 2016-09-29 at 16 49 07

**Comment by michalmarcinkowski** [comment]
Thank you Ania! 👍

**Pull request merged**